scionlab/models/user_as.py

```diff
@@ -187,72 +128,117 @@ class Meta:
+        Attach a UserAS to the specified attachment point creating the 
+        Link and the Interfaces, and if needed a VPNClient
+        :return Link link: The entity representing the relationship (UserAS, AttachmentPoint)
         """
-        prev_ap = self.attachment_point
-
-        host = self.hosts.get()   # UserAS always has only one host
-
+        # XXX: Should we check if the isd is consistent?
         if self.isd != attachment_point.AS.isd:
             self._change_isd(attachment_point.AS.isd)
-
-        host.update(
-            public_ip=public_ip,
-            bind_ip=_VAGRANT_VM_LOCAL_IP if installation_type == UserAS.VM else bind_ip,
-        )
-
-        link = self._get_ap_link()
-        interface_ap = link.interfaceA
-        interface_user = link.interfaceB
         if use_vpn:
-            vpn_client = self._create_or_activate_vpn_client(attachment_point.vpn)
-            interface_user.update(
-                public_ip=vpn_client.ip,
+            vpn_client = attachment_point.vpn.create_client(host=self.host, active=True)
+            interface_client = Interface.objects.create(
+                border_router=BorderRouter.objects.first_or_create(self.host),
+                vpn_client=vpn_client,
                 public_port=public_port,
-                bind_port=None
             )
-            interface_ap.update(
+            interface_ap = Interface.objects.create(
                 border_router=attachment_point.get_border_router_for_useras_interface(),
-                public_ip=attachment_point.vpn.server_vpn_ip,
-                public_port=None,
+                public_ip=attachment_point.vpn.server_vpn_ip
             )
         else:
-            host.vpn_clients.update(active=False)   # deactivate all vpn clients
-            interface_user.update(
-                public_ip=None,
+            interface_client = Interface.objects.create(
+                border_router=BorderRouter.objects.first_or_create(self.host),
+                public_ip=public_ip,
                 public_port=public_port,
-                bind_ip=None,
+                bind_ip=_VAGRANT_VM_LOCAL_IP if self.installation_type == UserAS.VM else bind_ip,
                 bind_port=bind_port
             )
-            interface_ap.update(
+            interface_ap = Interface.objects.create(
                 border_router=attachment_point.get_border_router_for_useras_interface(),
-                public_ip=None,
-                public_port=None,
             )
-        self.attachment_point = attachment_point
-        self.installation_type = installation_type
-        self.public_ip = public_ip
-        self.bind_ip = bind_ip
-        self.bind_port = bind_port
+
+        link = Link.objects.create(
+            type=Link.PROVIDER,
+            interfaceA=interface_ap,
+            interfaceB=interface_client
+        )
+        attachment_point.split_border_routers()
+        attachment_point.trigger_deployment()
```

Comment:
To avoid calling multiple times the `split_border_router(...)` and `trigger_deployment(...)` on the same `AttachmentPoint` we can use a proxy function which gets a list of `AttachmentPoint`s to be updated and updates the various connections toward an `AttachmentPoint` in batch, and then calls the two methods on it.
As far as the creation is concerned I don't think it would make much sense to have such a proxy, since right now the user can add only one `AttachmentPoint` at a time (and I believe this the most convenient way to go in order to keep everything simple and clean for the end user)

Reply:
Aha, I see. I had assumed user would be able to remove additional links (at least remove them in the UI, however that is handled internally), probably except for the last one so that there is always at least one link. 

Anyway, what I actually meant in this context with "removing a link" was changing the attachment-point which in a sense removes a link and creates a new one (even though that's not how this is (currently) handled internally). When changing from one AP to another one, both APs need to be `trigger_deployment`ed and `split_border_router`ed.